File: bindings/go/examples_readme_test.go
// Copyright The Pit Project Owners. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Please see https://github.com/openpitkit and the OWNERS file for details.

package openpit

import (
	"testing"
	"time"

	"go.openpit.dev/openpit/model"
	"go.openpit.dev/openpit/param"
	"go.openpit.dev/openpit/pkg/optional"
	"go.openpit.dev/openpit/pretrade/policies"
)

// Source: bindings/go/README.md - Usage
func TestReadmeQuickstart(t *testing.T) {
	usd, err := param.NewAsset("USD")
	if err != nil {
		t.Fatalf("NewAsset(USD) error = %v", err)
	}

	lowerBound, err := param.NewPnlFromString("-1000")
	if err != nil {
		t.Fatalf("NewPnlFromString(-1000) error = %v", err)
	}
	maxQty, err := param.NewQuantityFromString("500")
	if err != nil {
		t.Fatalf("NewQuantityFromString() error = %v", err)
	}
	maxNotional, err := param.NewVolumeFromString("100000")
	if err != nil {
		t.Fatalf("NewVolumeFromString() error = %v", err)
	}

	// 1. Configure and build the engine.
	engine, err := NewEngineBuilder().
		FullSync().
		Builtin(policies.BuildOrderValidation()).
		Builtin(
			policies.BuildPnlBoundsKillswitch().BrokerBarriers(
				policies.PnlBoundsBrokerBarrier{
					SettlementAsset: usd,
					LowerBound:      optional.Some(lowerBound),
				},
			),
		).
		Builtin(
			policies.BuildRateLimit().BrokerBarrier(
				policies.RateLimitBrokerBarrier{
					Limit: policies.RateLimit{MaxOrders: 100, Window: time.Second},
				},
			),
		).
		Builtin(
			policies.BuildOrderSizeLimit().
				BrokerBarrier(
					policies.OrderSizeBrokerBarrier{
						Limit: policies.OrderSizeLimit{
							MaxQuantity: maxQty,
							MaxNotional: maxNotional,
						},
					},
				).
				AssetBarriers(
					policies.OrderSizeAssetBarrier{
						SettlementAsset: usd,
						Limit: policies.OrderSizeLimit{
							MaxQuantity: maxQty,
							MaxNotional: maxNotional,
						},
					},
				),
		).
		Build()
	if err != nil {
		t.Fatalf("Build() error = %v", err)
	}
	defer engine.Stop()

	// 2. Check an order.
	order := model.NewOrder()
	op := order.EnsureOperationView()
	aapl, err := param.NewAsset("AAPL")
	if err != nil {
		t.Fatalf("NewAsset(AAPL) error = %v", err)
	}
	op.SetInstrument(param.NewInstrument(aapl, usd))
	op.SetAccountID(param.NewAccountIDFromInt(99224416))
	op.SetSide(param.SideBuy)
	price, _ := param.NewPriceFromString("185")
	qty, _ := param.NewQuantityFromString("100")
	op.SetTradeAmount(param.NewQuantityTradeAmount(qty))
	op.SetPrice(price)

	request, rejects, err := engine.StartPreTrade(order)
	if err != nil {
		t.Fatalf("StartPreTrade() error = %v", err)
	}
	if rejects != nil {
		t.Fatalf("StartPreTrade() unexpected rejects: %v", rejects)
	}
	defer request.Close()

	// 3. Real pre-trade and risk control.
	reservation, rejects, err := request.Execute()
	if err != nil {
		t.Fatalf("Execute() error = %v", err)
	}
	if rejects != nil {
		t.Fatalf("Execute() unexpected rejects: %v", rejects)
	}
	defer reservation.Close()

	// 4. Commit the reservation.
	reservation.Commit()

	// 5. Apply execution report.
	report := model.NewExecutionReport()
	reportOp := model.NewExecutionReportOperation()
	reportOp.SetInstrument(param.NewInstrument(aapl, usd))
	reportOp.SetAccountID(param.NewAccountIDFromInt(99224416))
	reportOp.SetSide(param.SideBuy)
	report.SetOperation(reportOp)

	pnl, _ := param.NewPnlFromString("-50")
	fee, _ := param.NewFeeFromString("3.4")
	impact := model.NewExecutionReportFinancialImpact()
	impact.SetPnl(pnl)
	impact.SetFee(fee)
	report.SetFinancialImpact(impact)

	result, err := engine.ApplyExecutionReport(report)
	if err != nil {
		t.Fatalf("ApplyExecutionReport() error = %v", err)
	}

	// 6. Kill switch must not be triggered after a small loss.
	if result.KillSwitchTriggered {
		t.Fatal("KillSwitchTriggered = true, want false")
	}
}

File: bindings/cpp/test/examples_readme_test.cpp
// Copyright The Pit Project Owners. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Please see https://openpit.dev and the OWNERS file for details.

// Source: bindings/cpp/README.md - Usage. Keep this code in sync with the
// public README snippet.

#include <openpit/openpit.hpp>

#include <stdexcept>
#include <string>

int main() {
  namespace model = openpit::model;
  namespace policies = openpit::pretrade::policies;

  openpit::EngineBuilder builder(openpit::SyncPolicy::None);
  builder.Add(policies::OrderValidationPolicy{});
  openpit::Engine engine = builder.Build();

  model::Order order = model::Order::Limit(
      model::Instrument("AAPL", "USD"),
      openpit::param::AccountId::FromUint64(99224416), model::Side::Buy,
      model::TradeAmount::OfQuantity(
          openpit::param::Quantity::FromString("100")),
      openpit::param::Price::FromString("185"));

  openpit::pretrade::StartResult start = engine.StartPreTrade(order);
  if (!start.Passed()) {
    const std::string reason = start.rejects.empty()
                                   ? "pre-trade start rejected"
                                   : start.rejects.front().reason;
    throw std::runtime_error(reason);
  }

  openpit::pretrade::ExecuteResult execute = start.request->Execute();
  if (!execute.Passed()) {
    const std::string reason = execute.rejects.empty()
                                   ? "pre-trade execute rejected"
                                   : execute.rejects.front().reason;
    throw std::runtime_error(reason);
  }

  execute.reservation->Commit();
  return 0;
}
